File: 50 Coding Challenge Part II/JavaScript/Challenge48.js
// Challenge 48: Create a function to return the longest
// word(s) in a string

function getTheLongestWord(text){
    let words = getWords(text);

    let maxSize = 0;
    let maxPositon = [];

    for (let i = 0; i < words.length; i++){
        let currWordSize = words[i].length;

        if (currWordSize > maxSize){
            maxSize = currWordSize;
            maxPositon = [ i ];
        }
        else if (currWordSize === maxSize){
            maxPositon.push(i)
        }
    }

    return getElements(words, maxPositon);
}

// Gets ily the elements frm specified positions from the array
function getElements(arr, arPositions){
    let arNew = [];

    for (let pos of arPositions){
        arNew.push(arr[pos]);
    }

    return arNew;
}


// Returns an array with the words from specified text
function getWords(text){
    let startWord = -1
    let arr = []

    for (let i = 0; i <= text.length; i++){
        let c = i < text.length ? text[i] : " ";

        if (!isSeperator(c) && startWord < 0){
            startWord = i;
        }

        if (isSeperator(c) && startWord >= 0){
            let word = text.substring(startWord, i);
            arr.push(word)

            startWord = -1;
        }
    }

    return arr;
}


// checks for seperators
function isSeperator(c){
    let seperators = [" ", "\t", "\n", "\r", ",", ";", ".", "!", "?", "(", ")"];
    return seperators.includes(c);
}

// Driver Method
let text = "Create a function to return the longest word(s) in a string"
console.log(getTheLongestWord(text));

File: 50 Coding Challenge Part II/Python/Challenge48.py
# Challenge 48: Create a function to return the longest word(s) in a string
def getTheLongestWord(text: str):
    words = getWords(text)
    maxSize = 0
    longestWords = []

    for word in words:
        currWordSize = len(word)
        if currWordSize > maxSize:
            maxSize = currWordSize
            longestWords = [word]
        elif currWordSize == maxSize:
            longestWords.append(word)
    return longestWords


def getWords(text: str):
    arrNew = []
    words = text.split(" ")
    for word in words:
        newAr = []
        for char in word:
            if not isSeperator(char):
                newAr.append(char)
        arrNew.append("".join(newAr))
    return arrNew


def isSeperator(char: str) -> bool:
    seperators = [" ", "\t", "\n", "\r", ",", ";", ".", "!", "?", "(", ")"]
    return char in seperators


# Driver Method
def main():
    text = "Create a function to return the longest word(s) in a string"
    print(getTheLongestWord(text))


if __name__ == "__main__":
    main()
